akka-projection-r2dbc/src/main/scala/akka/projection/r2dbc/internal/R2dbcOffsetStore.scala

```diff
@@ -135,29 +168,23 @@ private[projection] object R2dbcOffsetStore {
       }
     }
 
-    def window: JDuration =
-      JDuration.between(oldestTimestamp, latestTimestamp)
-
-    private lazy val sortedByTimestamp: Vector[Record] = byPid.valuesIterator.toVector.sortBy(_.timestamp)
-
-    lazy val latestBySlice: Vector[Record] = {
-      val builder = scala.collection.mutable.Map[Int, Record]()
-      sortedByTimestamp.reverseIterator.foreach { record =>
-        if (!builder.contains(record.slice))
-          builder.update(record.slice, record)
-      }
-      builder.values.toVector
-    }
-
-    def evict(until: Instant, keepNumberOfEntries: Int): State = {
-      if (oldestTimestamp.isBefore(until) && size > keepNumberOfEntries) {
-        val newState = State(
-          sortedByTimestamp.take(size - keepNumberOfEntries).filterNot(_.timestamp.isBefore(until))
-          ++ sortedByTimestamp.takeRight(keepNumberOfEntries)
-          ++ latestBySlice)
-        newState.copy(sizeAfterEvict = newState.size)
-      } else
+    def evict(slice: Int, timeWindow: JDuration): State = {
+      val recordsSortedByTimestamp = bySliceSorted.getOrElse(slice, TreeSet.empty[Record])
+      if (recordsSortedByTimestamp.isEmpty) {
         this
+      } else {
+        // this will always keep at least one, latest per slice
+        val until = recordsSortedByTimestamp.last.timestamp.minus(timeWindow)
```

Comment:
This idea will not fly. After scaling it may start at an earlier offset from some other slice than what is evicted/deleted .

 Good news is that I see where the problems are coming from now. I think we should change to the lazy loading approach of offsets as we did for dynamodb, and decouple deletes from the time window.